src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,206 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            linkEmptyList(value);
+            return;
+        }
+        Node<T> newNode = new Node<>(tail, value);
+        newNode.prev.next = newNode;
+        tail = newNode;
+        newNode.prev = tail.prev;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == 0 && size == 0) {
+            linkEmptyList(value);
+            return;
+        }
+        if (index == 0) {
+            linkHead(value);
+            return;
+        }
+        if (index == size) {
+            linkTail(value);
+            return;
+        }
+        checkIndexOutOfBounds(index);
+        Node<T> current = getNode(index);
+        Node<T> newNode = new Node<>(current.prev,value,current);
+        current.prev.next = newNode;
+        current.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list.size() == 0) {
+            return;
+        }
+        for (T element:list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexOutOfBounds(index);
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexOutOfBounds(index);
+        final Node<T> current = getNode(index);
+        final T oldValue = current.value;
+        current.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexOutOfBounds(index);
+        return unlink(getNode(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> x = head; x != null; x = x.next) {
+                if (x.value == null) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> x = head; x != null; x = x.next) {
+                if (object.equals(x.value)) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        }
         return false;
```

Comment:
combine these two blocks. You have the same body, same loop. Combine logical conditions